File: crm-app/frontend/src/utils/mapLinks.js
/**
 * Map link helpers — Google Maps must use exact saved pin coordinates only.
 */

import { getShopCompanyLabel } from './shopPinLabel';

const NEPAL_LAT = [26, 31];
const NEPAL_LNG = [80, 89];

function inRange(value, [min, max]) {
  return value >= min && value <= max;
}

function formatCoord(value) {
  return Number(value).toFixed(7);
}

/** Encode label for Google Maps query strings (spaces as +). */
function encodeMapLabel(label) {
  return encodeURIComponent(String(label).trim()).replace(/%20/g, '+');
}

/** "lat,lng+(Name)" — pins at exact coords; Google shows Name on the marker when supported. */
function coordsWithLabel(lat, lng, label) {
  const name = encodeMapLabel(label);
  return `${formatCoord(lat)},${formatCoord(lng)}+(${name})`;
}

/**
 * Resolve WGS84 pin coordinates. Returns null unless valid Nepal-range pin exists.
 */
export function resolveShopCoords(shop = {}) {
  let lat = Number(shop.latitude);
  let lng = Number(shop.longitude);

  if (!Number.isFinite(lat) || !Number.isFinite(lng)) return null;
  if (lat === 0 && lng === 0) return null;

  // Fix swapped lat/lng (common bug)
  if (inRange(lat, NEPAL_LNG) && inRange(lng, NEPAL_LAT)) {
    [lat, lng] = [lng, lat];
  }

  if (!inRange(lat, NEPAL_LAT) || !inRange(lng, NEPAL_LNG)) {
    return null;
  }

  return { lat, lng };
}

export function hasPinCoordinates(shop = {}) {
  return resolveShopCoords(shop) != null;
}

/** @deprecated Use hasPinCoordinates — directions require a saved pin. */
export function hasDirectionsTarget(shop = {}) {
  return hasPinCoordinates(shop);
}

export function googleMapsDestinationLabel(shop = {}) {
  return getShopCompanyLabel(shop);
}

/**
 * Google Maps driving directions to the exact saved pin with company name on the pin.
 * Uses saddr/daddr (not api=1) so coords stay exact and the label appears as "(Company)".
 * Avoid "Name@lat,lng" — that geocodes the name and can land on the wrong place.
 */
export function googleMapsDirectionsUrl(shop = {}, origin = null) {
  const dest = resolveShopCoords(shop);
  if (!dest) return null;

  const label = getShopCompanyLabel(shop);
  const daddr = coordsWithLabel(dest.lat, dest.lng, label);
  const from = origin
    ? resolveShopCoords({ latitude: origin.latitude, longitude: origin.longitude })
    : null;

  if (from) {
    const saddr = `${formatCoord(from.lat)},${formatCoord(from.lng)}`;
    return `https://www.google.com/maps?saddr=${saddr}&daddr=${daddr}&dir_action=navigate`;
  }

  return `https://www.google.com/maps?daddr=${daddr}&dir_action=navigate`;
}

/** Open Google Maps at exact pin with company name on the marker. */
export function googleMapsViewUrl(shop = {}) {
  const coords = resolveShopCoords(shop);
  if (!coords) return null;
  const q = coordsWithLabel(coords.lat, coords.lng, getShopCompanyLabel(shop));
  return `https://www.google.com/maps?q=${q}`;
}

export function openGoogleMapsDirectionsFromHere(shopArgs, { onStatus, onError, onFinally } = {}) {
  const dest = resolveShopCoords(shopArgs);
  if (!dest) {
    onError?.('No map pin saved. Open the shop, place the red pin on the map, and save.');
    onFinally?.();
    return;
  }

  const open = (origin) => {
    const url = googleMapsDirectionsUrl(shopArgs, origin);
    if (!url) {
      onError?.('Could not build directions link.');
      onFinally?.();
      return;
    }
    window.open(url, '_blank', 'noopener,noreferrer');
    onFinally?.();
  };

  if (!navigator.geolocation) {
    open(null);
    return;
  }

  onStatus?.('Getting location…');

  navigator.geolocation.getCurrentPosition(
    (position) => {
      open({
        latitude: position.coords.latitude,
        longitude: position.coords.longitude,
      });
    },
    () => open(null),
    { enableHighAccuracy: true, maximumAge: 0, timeout: 20000 }
  );
}

File: crm-mobile/shared/src/commonMain/kotlin/com/gurucrm/mobile/util/MapLinks.kt
package com.gurucrm.mobile.util

import com.gurucrm.mobile.data.LatLng
import com.gurucrm.mobile.format

private val nepalLat = 26.0..31.0
private val nepalLng = 80.0..89.0

fun resolveShopCoords(latitude: Double?, longitude: Double?): LatLng? {
    var lat = latitude ?: return null
    var lng = longitude ?: return null
    if (lat == 0.0 && lng == 0.0) return null
    if (lat in nepalLng && lng in nepalLat) {
        val tmp = lat
        lat = lng
        lng = tmp
    }
    if (lat !in nepalLat || lng !in nepalLng) return null
    return LatLng(lat, lng)
}

private fun formatCoord(value: Double): String = value.format(7)

private fun encodeMapLabel(label: String): String =
    encodeURIComponent(label.trim()).replace("%20", "+")

private fun encodeURIComponent(value: String): String =
    value.encodeToByteArray().joinToString("") { byte ->
        val c = byte.toInt() and 0xFF
        when {
            c in 'a'.code..'z'.code || c in 'A'.code..'Z'.code || c in '0'.code..'9'.code ||
                c == '-'.code || c == '_'.code || c == '.'.code || c == '~'.code -> c.toChar().toString()
            c == ' '.code -> "+"
            else -> "%%${c.toString(16).uppercase().padStart(2, '0')}"
        }
    }

private fun coordsWithLabel(lat: Double, lng: Double, label: String): String {
    val name = encodeMapLabel(label)
    return "${formatCoord(lat)},${formatCoord(lng)}+($name)"
}

fun googleMapsDirectionsUrl(
    destLat: Double,
    destLng: Double,
    shopName: String,
    origin: LatLng? = null,
): String {
    val daddr = coordsWithLabel(destLat, destLng, shopName)
    if (origin != null) {
        val saddr = "${formatCoord(origin.lat)},${formatCoord(origin.lng)}"
        return "https://www.google.com/maps?saddr=$saddr&daddr=$daddr&dir_action=navigate"
    }
    return "https://www.google.com/maps?daddr=$daddr&dir_action=navigate"
}

fun formatRouteDistance(meters: Double): String {
    if (meters < 1000) return "${meters.toInt()} m"
    return "${(meters / 1000).format(1)} km"
}

fun formatRouteDuration(seconds: Double): String {
    val mins = (seconds / 60).toInt()
    if (mins < 60) return "$mins min"
    val hours = mins / 60
    val rem = mins % 60
    return if (rem == 0) "${hours}h" else "${hours}h ${rem}m"
}
